### PR TITLE
Use SolrProcessor.get_pub_year to get edition years

### LINKED ISSUE #6021: Date format isn't correctly read off edition.
### Evidence / Screenshot (if possible)
Screenshot_2021-12-29 Kerstin Gier Open Library

### Relevant url?

[Affected work](https://openlibrary.org/works/OL26745647W)

### Steps to Reproduce

1. Go to ...
2. Do ...

* Actual:
* Expected:

### Details

when adding the publication date to editions in the recommended format (YYYY-MM-DD), the website isn't able to display the line "First published in YYYY" on an author's page. "Sep 29th, 2021" would work on the other hand. This doesn't allow to correctly sort by most recent/first published.

### Proposal & Constraints
as both formats "YYYY" and "Sep 29th, 2021" work and are correctly sorted on a work's page, I can imagine those two are converted to the same format in order to be able to be compared to each other. If that's not the case for "YYYY-MM-DD", it would be worth investigating if that effects the sort order in the edition list as well.

### Stakeholders
@cdrini

### PR DESCRIPTION
**Pull request by agmckee**
get_pub_year already supports ISO dates, as well as simple year strings, so it is an easy fix for the issue described in #6021. I don't believe this can have any negative side effects, it will simply extend support to ISO 8601 formatted dates without affecting dates currently indexed normally as get_pub_year falls back to re_year.search if re_iso_date doesn't match.

Closes #6021

fix

### Technical

This is a fix as currently the document indexed by SOLR for works with only ISO dates doesn't have the "first_published_year" attribute, causing the UI issue described in #6021.

### Testing

Add a work and edition with an ISO 8601 format date and look at the SOLR index for that work, you should notice the lack of "publish_year" and "first_publish_year" attributes. This affects worksearch, and the UI as described in #6021. After checking out the changes in this PR please reindex `docker-compose exec web make reindex-solr` and you should now see the aforementioned attributes for the respective work in SOLR. 

### Screenshot
Example of record with ISO 8601 formatted publish date.

example-record

SearchResultsWork macro displaying the year correctly.

example-searchresultswork-macro

### Stakeholders

@mekarpeles @cdrini

### CODE REVIEW

File: openlibrary/tests/solr/test_update_work.py [hunk 1]
```diff
@@ -164,7 +164,7 @@ def test_publish_year(self):
         test_dates = [
             "2000",
             "Another 2000",
-            "2001-01-02",  # Doesn't seems to be handling this case
+            "2001-01-02",  # ISO 8601 formmated dates now supported
```
Comment (agmckee) [inline openlibrary/tests/solr/test_update_work.py:167]: I'll correct this typo in a new commit tomorrow, sorry about that.

**Comment by agmckee** [comment]
Tests initially failed because (a) the tests currently have an ISO 8601 formatted date in them that is noted as not being covered by current functionality and (b) first version of code didn't handle the None values that sometimes arise, this is now rectified.
The pre-commit check appears to be failing for lots of PRs at present and is, I believe, unrelated.

**Comment by jimman2003** [comment]
Yeah, another (unrelated) part of the code wasn't updated to the new syntax, hence the test fail

**Review by cdrini — APPROVED** [review]
Awesome thank you @agmckee ! And thank you @jimman2003 for helping review :) 

I updated this to be a bit more lenient and also take years at the front since we often have them in the form eg `2004 May 19`. The tests look not unreliable here, so not testing manually. Lgtm!

**Comment by cdrini** [comment]
This has been merged into master, and any newly edited works will have their dates corrected! Once we do the next full reindex (#5827), all records will be updated and run with the new code here!

**Pull request merged**